nx/lib/nx/serving.ex

```diff
@@ -176,17 +252,329 @@ defmodule Nx.Serving do
-                "Got: #{inspect(other)}"
+  @doc false
+  def child_spec(opts) when is_list(opts) do
+    name = opts[:name]
+
+    if name == nil or not is_atom(name) do
+      raise ArgumentError, ":name option is expected when starting Nx.Serving and must be an atom"
     end
+
+    opts[:serving] ||
+      raise ArgumentError, ":serving option is expected when starting Nx.Serving"
+
+    %{
+      id: name,
+      start: {__MODULE__, :start_link, [opts]},
+      type: :supervisor
+    }
+  end
+
+  @doc """
+  TODO
+  """
+  def start_link(opts) do
+    children = [
+      %{
+        id: __MODULE__,
+        start: {__MODULE__, :start_child, [opts]}
+      }
+    ]
+
+    Supervisor.start_link(children, strategy: :one_for_all, max_restarts: 0)
+  end
+
+  @doc """
+  TODO
+  """
+  def batched_run(name, input) when is_atom(name) do
+    pid = GenServer.whereis(name) || exit({:noproc, {__MODULE__, :batched_run, [name, input]}})
+
+    %{
+      preprocessing: preprocessing,
+      postprocessing: postprocessing,
+      limit: limit
+    } = :persistent_term.get(persistent_key(name))
+
+    {batch, info} = handle_preprocessing(preprocessing, input)
+
+    if batch.size > limit do
+      raise ArgumentError,
+            "batch size (#{batch.size}) cannot exceed Nx.Serving server batch size of #{limit}"
+    end
+
+    # Use Process.monitor/2 on Elixir v1.15+
+    ref = :erlang.monitor(:process, pid, alias: :demonitor)
+    Process.send(pid, {__MODULE__, :batched_run, ref, batch}, [:noconnect])
+
+    {tensor, metadata} = receive_batched(batch.size, ref, [], nil, name, input)
+    handle_postprocessing(postprocessing, tensor, metadata, info)
+  end
+
+  @axis 0
+
+  defp receive_batched(0, ref, acc, {template, metadata}, _name, _input) do
+    Process.demonitor(ref, [:flush])
+
+    tensors =
+      acc
+      |> Enum.reverse()
+      |> Enum.zip_with(&Nx.concatenate(&1, axis: @axis))
+
+    {output, []} =
+      Nx.Defn.Composite.traverse(template, tensors, fn _template, [tensor | tensors] ->
+        {tensor, tensors}
+      end)
+
+    {output, metadata}
+  end
+
+  defp receive_batched(total_size, ref, acc, _template_metadata, name, input) do
+    receive do
+      {^ref, {start, size, output, metadata}} ->
+        # If we have a single response, slice and return immediately.
+        # Otherwise we collect their contents and build the concatenated result later.
+        if acc == [] and size == total_size do
+          Process.demonitor(ref, [:flush])
+
+          output =
+            Nx.Defn.Composite.traverse(output, &Nx.slice_along_axis(&1, start, size, axis: @axis))
+
+          {output, metadata}
+        else
+          funs =
+            output
+            |> Nx.Defn.Composite.reduce(
+              [],
+              &[Nx.slice_along_axis(&1, start, size, axis: @axis) | &2]
+            )
+            |> Enum.reverse()
+
+          receive_batched(total_size - size, ref, [funs | acc], {output, metadata}, name, input)
```

Comment:
We use the first `receive_batched` clause only here, and it doesn't actually receive. We could check if `total_size - size` is 0 here and call either `receive_batched` or `finalize_batched`?

Reply:
I tried but it only added more nesting :( so I keep the current version for now. If you have ideas, feel free to PR. :)

Reply:
I ended up with this, not sure if it's better, but the flow is probably easier to follow. Whatever you prefer!

```diff
diff --git a/nx/lib/nx/serving.ex b/nx/lib/nx/serving.ex
index 84f58e3..b87b362 100644
--- a/nx/lib/nx/serving.ex
+++ b/nx/lib/nx/serving.ex
@@ -315,50 +315,24 @@ defmodule Nx.Serving do
     ref = :erlang.monitor(:process, pid, alias: :demonitor)
     Process.send(pid, {__MODULE__, :batched_run, ref, batch}, [:noconnect])
 
-    {tensor, metadata} = receive_batched(batch.size, ref, [], nil, name, input)
+    {tensor, metadata} = receive_batched(batch.size, ref, [], name, input)
     handle_postprocessing(postprocessing, tensor, metadata, info)
   end
 
   @axis 0
 
-  defp receive_batched(0, ref, acc, {template, metadata}, _name, _input) do
-    Process.demonitor(ref, [:flush])
-
-    tensors =
-      acc
-      |> Enum.reverse()
-      |> Enum.zip_with(&Nx.concatenate(&1, axis: @axis))
-
-    {output, []} =
-      Nx.Defn.Composite.traverse(template, tensors, fn _template, [tensor | tensors] ->
-        {tensor, tensors}
-      end)
-
-    {output, metadata}
-  end
-
-  defp receive_batched(total_size, ref, acc, _template_metadata, name, input) do
+  defp receive_batched(total_size, ref, acc, name, input) do
     receive do
       {^ref, {start, size, output, metadata}} ->
-        # If we have a single response, slice and return immediately.
-        # Otherwise we collect their contents and build the concatenated result later.
-        if acc == [] and size == total_size do
-          Process.demonitor(ref, [:flush])
-
-          output =
-            Nx.Defn.Composite.traverse(output, &Nx.slice_along_axis(&1, start, size, axis: @axis))
-
-          {output, metadata}
-        else
-          funs =
-            output
-            |> Nx.Defn.Composite.reduce(
-              [],
-              &[Nx.slice_along_axis(&1, start, size, axis: @axis) | &2]
-            )
-            |> Enum.reverse()
-
-          receive_batched(total_size - size, ref, [funs | acc], {output, metadata}, name, input)
+        complete? = total_size == size
+
+        case handle_batched_response(start, size, output, complete?, acc) do
+          {:cont, acc} ->
+            receive_batched(total_size - size, ref, acc, name, input)
+
+          {:halt, tensor} ->
+            Process.demonitor(ref, [:flush])
+            {tensor, metadata}
         end
 
       {:DOWN, ^ref, _, _, reason} ->
@@ -368,6 +342,43 @@ defmodule Nx.Serving do
     end
   end
 
+  # If we have a single response, slice and return immediately.
+  defp handle_batched_response(start, size, output, _complete? = true, _acc = []) do
+    tensor =
+      Nx.Defn.Composite.traverse(output, &Nx.slice_along_axis(&1, start, size, axis: @axis))
+
+    {:halt, tensor}
+  end
+
+  # Otherwise we collect their contents and build the concatenated result later.
+  defp handle_batched_response(start, size, output, complete?, acc) do
+    chunks =
+      output
+      |> Nx.Defn.Composite.reduce(
+        [],
+        &[Nx.slice_along_axis(&1, start, size, axis: @axis) | &2]
+      )
+      |> Enum.reverse()
+
+    acc = [chunks | acc]
+
+    if complete? do
+      tensors =
+        acc
+        |> Enum.reverse()
+        |> Enum.zip_with(&Nx.concatenate(&1, axis: @axis))
+
+      {tensor, []} =
+        Nx.Defn.Composite.traverse(template, tensors, fn _template, [tensor | tensors] ->
+          {tensor, tensors}
+        end)
+
+      {:halt, tensor}
+    else
+      {:cont, acc}
+    end
+  end
+
   ## Process callbacks
 
   require Logger
```